### Анализ базы данных сервиса книг

**Цель**: Подготовка предложений для развития нового сервиса для чтения книг

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### План работы:

- Посчитаем, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитаем количество обзоров и среднюю оценку;
- Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками;
- Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [1]:
# импортируем библиотеки

import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры

In [2]:
# Напишем функцию для вывода 5 строк каждой таблицы

def open_table (table_name):
    return pd.io.sql.read_sql("SELECT * FROM {} ".format(table_name), con = engine)  
for i in ['books', 'authors','publishers','ratings', 'reviews']:
    print(i)
    display(open_table(i))
    display(open_table(i).info())

books


book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

authors


author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

publishers


publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

ratings


rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

reviews


review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

In [3]:
# Сколько книг вышло после 1 января 2000 года

query = """SELECT COUNT(*)
FROM books
WHERE publication_date > DATE '2000-01-01'"""

pd.io.sql.read_sql(query,con = engine)

count
0    819

С начала 2000 года в базе данных есть информация о выходе 819 книг.

In [4]:
# Для каждой книги посчитаем количество обзоров и среднюю оценку

query = """ WITH avg_rating as 
(SELECT ratings.book_id, AVG(ratings.rating) as avg_rating FROM ratings GROUP BY ratings.book_id), 
cnt_review as (SELECT reviews.book_id, COUNT(reviews.review_id) as cnt_review FROM reviews GROUP BY reviews.book_id)
SELECT title, avg_rating, cnt_review
FROM books
JOIN avg_rating ON books.book_id = avg_rating.book_id
JOIN cnt_review ON books.book_id = cnt_review.book_id
ORDER BY avg_rating DESC, cnt_review DESC
"""
pd.io.sql.read_sql(query,con = engine)

title  avg_rating  cnt_review
0                        A Dirty Job (Grim Reaper  #1)        5.00           4
1         Moneyball: The Art of Winning an Unfair Game        5.00           3
2              School's Out—Forever (Maximum Ride  #2)        5.00           3
3                             Neil Gaiman's Neverwhere        5.00           2
4             Evil Under the Sun (Hercule Poirot  #24)        5.00           2
..                                                 ...         ...         ...
989  The World Is Flat: A Brief History of the Twen...        2.25           3
990                                      Drowning Ruth        2.00           3
991                  His Excellency: George Washington        2.00           2
992                                              Junky        2.00           2
993                               Harvesting the Heart        1.50           2

[994 rows x 3 columns]

Самая популярная книга - A Dirty Job (Grim Reaper #1).

In [5]:
# Определим издательство, которое выпустило наибольшее число книг толще 50 страниц

query = """SELECT publisher, COUNT(book_id) as cnt_book
FROM publishers
JOIN books on books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP BY publisher
ORDER BY cnt_book DESC
LIMIT 1"""
pd.io.sql.read_sql(query, con = engine)

publisher  cnt_book
0  Penguin Books        42

Больше всего книг выпустило издательство `Penguin Books`.

In [11]:
# Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками

query = """WITH avg_rating as 
    (SELECT ratings.book_id, AVG(ratings.rating) as avg_rating_book, COUNT(ratings.rating) as cnt_rating_book  
    FROM ratings 
    GROUP BY ratings.book_id)
SELECT author, COUNT(cnt_rating_book) as cnt_rating_author, AVG(avg_rating_book) as avg_rating_author
FROM books
JOIN avg_rating ON books.book_id = avg_rating.book_id
JOIN authors on books.author_id = authors.author_id
WHERE cnt_rating_book >=50
GROUP BY author
ORDER BY avg_rating_author DESC
"""
pd.io.sql.read_sql(query, con = engine)

author  cnt_rating_author  \
0                          J.K. Rowling/Mary GrandPré                  4   
1                   Markus Zusak/Cao Xuân Việt Khương                  1   
2                                      J.R.R. Tolkien                  2   
3                                   Louisa May Alcott                  1   
4                                        Rick Riordan                  1   
5                                     William Golding                  1   
6                                       J.D. Salinger                  1   
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce                  1   
8   William Shakespeare/Paul Werstine/Barbara A. M...                  1   
9                                           Dan Brown                  2   
10                                         Lois Lowry                  1   
11            George Orwell/Boris Grabnar/Peter Škerl                  1   
12                                    Stephenie Meyer                  1   
13                                     John Steinbeck                  1   

    avg_rating_author  
0            4.283844  
1            4.264151  
2            4.258446  
3            4.192308  
4            4.080645  
5            3.901408  
6            3.825581  
7            3.789474  
8            3.787879  
9            3.754540  
10           3.750000  
11           3.729730  
12           3.662500  
13           3.622951

Самая популярная писательница по оценкам пользователей, автор книг о Гарри Поттере - `Дж. Роулинг`.

In [7]:
# Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

query = """WITH 
    cnt_review as (SELECT reviews.username, COUNT(reviews.review_id) as cnt_reviews
    FROM reviews
    GROUP BY reviews.username), 
    
    cnt_rating as (SELECT ratings.username, COUNT(ratings.rating) as cnt_rating
    FROM ratings
    GROUP BY ratings.username
    HAVING COUNT(ratings.rating) >50)
    
SELECT AVG(cnt_review.cnt_reviews)
FROM cnt_rating
JOIN cnt_review ON cnt_review.username = cnt_rating.username"""
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

Среднее количество обзоров от пользователей с высоким количеством оценок - 24.

**Выводы:**
1. Большее число пользователей предпочитает книги Дж. Роулинг.
2. Возможно предусмотреть систему скидок или поощрения пользователей, которые оставляют больше всего оценок и пишут больше всего обзоров.